**Table of contents**<a id='toc0_'></a>    
- 1. [Problem 1: Production economy and CO2 taxation](#toc1_)    
- 2. [Problem 2: Career choice model](#toc2_)    
- 3. [Problem 3: Barycentric interpolation](#toc3_)    

<!-- vscode-jupyter-toc-config
	numbering=true
	anchor=true
	flat=false
	minLevel=2
	maxLevel=6
	/vscode-jupyter-toc-config -->
<!-- THIS CELL WILL BE REPLACED ON TOC UPDATE. DO NOT WRITE YOUR TEXT IN THIS CELL -->

In [1]:
import numpy as np
from types import SimpleNamespace
from scipy.optimize import minimize
from exam_2024_NoahPetra import MarketClearing
from exam_2024_NoahPetra import MarketClearing2
%load_ext autoreload
%autoreload 2

## 1. <a id='toc1_'></a>[Problem 1: Production economy and CO2 taxation](#toc0_)

Consider a production economy with two firms indexed by $j \in \{1,2\}$. Each produce its own good. They solve

$$
\begin{align*}
\max_{y_{j}}\pi_{j}&=p_{j}y_{j}-w_{j}\ell_{j}\\\text{s.t.}\;&y_{j}=A\ell_{j}^{\gamma}.
\end{align*}
$$

Optimal firm behavior is

$$
\begin{align*}
\ell_{j}^{\star}(w,p_{j})&=\left(\frac{p_{j}A\gamma}{w}\right)^{\frac{1}{1-\gamma}} \\
y_{j}^{\star}(w,p_{j})&=A\left(\ell_{j}^{\star}(w,p_{j})\right)^{\gamma}
\end{align*}
$$

The implied profits are

$$
\pi_{j}^*(w,p_{j})=\frac{1-\gamma}{\gamma}w\cdot\left(\frac{p_{j}A\gamma}{w}\right)^{\frac{1}{1-\gamma}}
$$

A single consumer supplies labor, and consumes the goods the firms produce. She also recieves the implied profits of the firm.<br>
She solves:

$$
\begin{align*}
U(p_1,p_2,w,\tau,T) = \max_{c_{1},c_{2},\ell} & \log(c_{1}^{\alpha}c_{2}^{1-\alpha})-\nu\frac{\ell^{1+\epsilon}}{1+\epsilon} \\
\text{s.t.}\,\,\,&p_{1}c_{1}+(p_{2}+\tau)c_{2}=w\ell+T+\pi_{1}^*(w,p_{1})+\pi_{2}^*(w,p_{2})
\end{align*}
$$

where $\tau$ is a tax and $T$ is lump-sum transfer. <br>
For a given $\ell$, it can be shown that optimal behavior is

$$
\begin{align*}
c_{1}(\ell)&=\alpha\frac{w\ell+T+\pi_{1}^*(w,p_{1})+\pi_{2}^*(w,p_{2})}{p_{1}} \\
c_{2}(\ell)&=(1-\alpha)\frac{w\ell+T+\pi_{1}^*(w,p_{1})+\pi_{2}^*(w,p_{2})}{p_{2}+\tau} \\
\end{align*}
$$
Such that optimal behavior is:
$$
\ell^* = \underset{\ell}{\arg\max} \log(\left(c_{1}(\ell)\right)^{\alpha}\cdot \left(c_{2}(\ell)\right)^{1-\alpha})-\nu\frac{\ell^{1+\epsilon}}{1+\epsilon} 
$$
With optimal consumption:
$$
\begin{align*}
c_1^*=c_{1}(\ell^*) \\
c_2^*=c_{2}(\ell^*)\\
\end{align*}
$$


The government chooses $\tau$ and balances its budget so $T=\tau c_2^*$. We initially set $\tau,T=0$.

Market clearing requires:

1. Labor market: $\ell^* = \ell_1^* + \ell_2^*$
1. Good market 1: $c_1^* = y_1^*$
1. Good market 2: $c_2^* = y_2^*$


**Question 1:** Check market clearing conditions for $p_1$ in `linspace(0.1,2.0,10)` and $p_2$ in `linspace(0.1,2.0,10)`. We choose $w=1$ as numeraire.

In [2]:
par = SimpleNamespace()

# firms
par.A = 1.0
par.gamma = 0.5

# households
par.alpha = 0.3
par.nu = 1.0
par.epsilon = 2.0

# government
par.tau = 0.0
par.T = 0.0

# Question 3
par.kappa = 0.1

w = 1.0  # Numeraire

In [3]:
# Instantiate the MarketClearing class
market = MarketClearing(par.A, par.gamma, par.alpha, par.nu, par.epsilon, par.tau, par.T, w)

# Check market clearing conditions
results = market.check_market_clearing(par)

# Print results
market.print_results(results)

p1: 0.10, p2: 0.10 | Labor Clearing: True | Good 1 Clearing: False | Good 2 Clearing: False
p1: 0.10, p2: 0.31 | Labor Clearing: True | Good 1 Clearing: False | Good 2 Clearing: False
p1: 0.10, p2: 0.52 | Labor Clearing: True | Good 1 Clearing: False | Good 2 Clearing: False
p1: 0.10, p2: 0.73 | Labor Clearing: True | Good 1 Clearing: False | Good 2 Clearing: False
p1: 0.10, p2: 0.94 | Labor Clearing: True | Good 1 Clearing: False | Good 2 Clearing: False
p1: 0.10, p2: 1.16 | Labor Clearing: True | Good 1 Clearing: False | Good 2 Clearing: False
p1: 0.10, p2: 1.37 | Labor Clearing: True | Good 1 Clearing: False | Good 2 Clearing: False
p1: 0.10, p2: 1.58 | Labor Clearing: True | Good 1 Clearing: False | Good 2 Clearing: False
p1: 0.10, p2: 1.79 | Labor Clearing: True | Good 1 Clearing: False | Good 2 Clearing: False
p1: 0.10, p2: 2.00 | Labor Clearing: True | Good 1 Clearing: False | Good 2 Clearing: False
p1: 0.31, p2: 0.10 | Labor Clearing: True | Good 1 Clearing: False | Good 2 Clea

Within the range specified by the linspace of the question, there is no combination of $p1$ and $p2$ that clears the market for good 1 or good 2. However, the labor market clears for every value in the specified range.

**Question 2:** Find the equilibrium prices $p_1$ and $p_2$.<br>
*Hint: you can use Walras' law to only check 2 of the market clearings*

In [4]:
# Instantiate the MarketClearing class
market = MarketClearing(par.A, par.gamma, par.alpha, par.nu, par.epsilon, par.tau, par.T, w)

# Solve for equilibrium prices
equilibrium_prices = market.solve_equilibrium(par)
print(f"Equilibrium Prices: p1 = {equilibrium_prices[0]}, p2 = {equilibrium_prices[1]}")

Equilibrium Prices: p1 = 0.7816943854085492, p2 = 1.194057897274884


In [5]:
# Instantiate the MarketClearing class
market = MarketClearing(par.A, par.gamma, par.alpha, par.nu, par.epsilon, par.tau, par.T, w)

# Solve for equilibrium prices
equilibrium_prices = market.solve_equilibrium(par)
print(f"Equilibrium Prices: p1 = {equilibrium_prices[0]}, p2 = {equilibrium_prices[1]}")

# Calculate equilibrium values at the obtained prices
equilibrium_values = market.calculate_equilibrium_values(equilibrium_prices, par)

# Print equilibrium values
print("Equilibrium Values:")
for key, value in equilibrium_values.items():
    print(f"{key}: {value}")

# Check market clearing
labor_clearing = equilibrium_values['labor_clearing']
good1_clearing = equilibrium_values['good1_clearing']
good2_clearing = equilibrium_values['good2_clearing']

Equilibrium Prices: p1 = 0.7816943854085492, p2 = 1.194057897274884
Equilibrium Values:
ell1_star: 0.15276152804481236
ell2_star: 0.35644356551112943
ell_star: 0.5092050935559418
y1_star: 0.3908471927042746
y2_star: 0.597028948637442
pi1_star: 0.15276152804481236
pi2_star: 0.35644356551112943
income: 1.0184101871118836
c1_star: 0.39084719276048624
c2_star: 0.5970289486006428
labor_clearing: True
good1_clearing: True
good2_clearing: True


The verification results confirm that the markets do clear with the obtained equilibrium prices:

- **Labor Market Clearing**: True
- **Good 1 Market Clearing**: True
- **Good 2 Market Clearing**: True

Here are the detailed intermediate values at the equilibrium prices:

1. **Optimal Labor Supply**:
   \[
   \begin{align*}
   \ell_1^* &\approx 0.153 \\
   \ell_2^* &\approx 0.356 \\
   \ell^* &\approx 0.509 \text{ (Total Labor)}
   \end{align*}
   \]

2. **Optimal Output**:
   \[
   \begin{align*}
   y_1^* &\approx 0.391 \\
   y_2^* &\approx 0.597
   \end{align*}
   \]

3. **Implied Profits**:
   \[
   \begin{align*}
   \pi_1^* &\approx 0.153 \\
   \pi_2^* &\approx 0.356
   \end{align*}
   \]

4. **Optimal Consumption**:
   \[
   \begin{align*}
   \text{Total Income} &\approx 1.018 \\
   c_1^* &\approx 0.391 \\
   c_2^* &\approx 0.597
   \end{align*}
   \]

### Conclusion

The equilibrium prices $p_1$ = 0.7816943854085492 and $p_2$ = 1.194057897274884 lead to market clearing for labor and both goods, confirming the correctness of the solution. This indicates that the model and parameters used can achieve equilibrium under these conditions.

Assume the government care about the social welfare function:

$$
SWF = U - \kappa y_2^*
$$

Here $\kappa$ measures the social cost of carbon emitted by the production of $y_2$ in equilibrium.

**Question 3:** What values of $\tau$ and (implied) $T$ should the government choose to maximize $SWF$?

In [6]:
# Initialize parameters
par = SimpleNamespace()
par.A = 1.0
par.gamma = 0.5
par.alpha = 0.3
par.nu = 1.0
par.epsilon = 2.0
par.w = 1.0    # Initial value for w
par.kappa = 0.1

# Create an instance of the MarketClearing class
market = MarketClearing2(par.A, par.gamma, par.alpha, par.nu, par.epsilon, 0.0, 0.0, par.w)

# Define equilibrium prices
equilibrium_prices = (0.7816943854085492, 1.194057897274884)

# Solve for the optimal tau and T with the constraint
optimal_tau_revised, optimal_T_revised = market.solve_continuous_with_constraint(equilibrium_prices)

# Calculate the SWF value using the new optimal tau and T
SWF_value_revised = -market.objective_SWF_with_constraint((optimal_tau_revised, optimal_T_revised))

print(f"SWF Value: {SWF_value_revised}")

Equilibrium prices: p1 = 0.7816943854085492, p2 = 1.194057897274884
Optimal tau: 0.1
Optimal T: 2.0
SWF Value: 0.28359187215663206
